### ISSUE TITLE
[dev.icinga.com #4199] multiple idomod modules: only first gets data from registered callback functions

### ISSUE THREAD

**Issue by icinga-migration** [Icinga/icinga-core#1282]
This issue has been migrated from Redmine: https://dev.icinga.com/issues/4199

**Created by mfriedrich on 2013-05-18 17:45:50 +00:00**

Assignee: _mfriedrich_
Status: _Resolved_  (closed on _2013-05-22 11:29:35 +00:00_)
Target Version: _1.10_
Last Update: _2013-05-22 11:29:35 +00:00 (in Redmine)_

    Icinga Version: 1.9.0
    OS Version: any

---

once a module gets loaded via dlopen() it registers its callback functions. if one loads multiple modules of the same binary, the exported symbol space is shared and only the first entry wins.

http://pubs.opengroup.org/onlinepubs/009695399/functions/dlopen.html

    Symbols introduced into a program through calls to dlopen() may be used in relocation activities. Symbols so introduced may duplicate symbols already defined by the program or previous dlopen() operations. To resolve the ambiguities such a situation might present, the resolution of a symbol reference to symbol definition is based on a symbol resolution order. Two such resolution orders are defined: load or dependency ordering. Load order establishes an ordering among symbol definitions, such that the definition first loaded (including definitions from the image file and any dependent objects loaded with it) has priority over objects added later (via dlopen()). Load ordering is used in relocation processing. Dependency ordering uses a breadth-first order starting with a given object, then all of its dependencies, then any dependents of those, iterating until all dependencies are satisfied. With the exception of the global symbol object obtained via a dlopen() operation on a file of 0, dependency ordering is used by the dlsym() function. Load ordering is used in dlsym() operations upon the global symbol object.

**Changesets**

_2013-05-18 18:51:17 +00:00_ by _(unknown)_ 524838659e715f1edc57f0ef96d8ab02544347ef

    nebmods: fix multiple modules sharing same symbols not receiving callback data
    
    once dlopen() loads modules with the same name, it will make sure to
    register only the first symbols, and not overriding the coming ones from
    more modules of the same binary.
    
    this is bad, as e.g. multiple idomod's will try to register the same
    callback function symbols, and only the first one wins then. this is the
    reason why only the first instance_name is getting populated in idoutils
    database then too.
    
    the initial handshake of ido2db with idomod only happens for the reason
    of calling the nebmodule_init() function symbol (from the first module)
    with different arguments, passed by the core. each module should have at
    least a different instance_name defined in its idomod.cfg, so the init
    function will be called with those, doing a handshake (and insert into
    icinga_instances table then) then. but afterwards, no data gets received
    from the callbacks not having such a diversity (how should the core then
    know anyways, it's a _callback_ function called only once from symbol
    space!).
    
    basically, the root issue is an omd patch living for more than 2 years
    now, but the old behaviour is now restored with this fix (plus adding a
    debug idea stolen from Andreas Ericsson, thanks).
    alongside fixing this, the solution is rather simple - make a temp copy
    of the binary to some random name, and let dlopen actually load that
    into memory. then all the registered function symbols will remain module
    based and everything else works with the callbacks and so on.
    
    as this is a behavioural change, it should get a note in Changelog once
    merged into release trees.
    
    refs #4199

_2013-05-22 12:46:13 +00:00_ by _(unknown)_ a84cd3d6b63b6de4bf4fae6114586b8a2e0ea206

    update Changelog for #4199
    
    refs #4199
